targets/TARGET_NUVOTON/TARGET_M2351/rtc_api.c

```diff
@@ -0,0 +1,246 @@
+/* POSIX time set by rtc_write() */
+static time_t t_write = 0;
+/* Convert date time from H/W RTC to struct TM */
+static void rtc_convert_datetime_hwrtc_to_tm(struct tm *datetime_tm, const S_RTC_TIME_DATA_T *datetime_hwrtc);
+
+static const struct nu_modinit_s rtc_modinit = {RTC_0, RTC_MODULE, 0, 0, 0, RTC_IRQn, NULL};
+
+__NONSECURE_ENTRY
+void rtc_init(void)
+{
+    if (rtc_isenabled()) {
+        return;
+    }
+
+    RTC_Open(NULL);
+
+    /* POSIX time origin (00:00:00 UTC, Thursday, 1 January 1970) */
+    rtc_write(0);
+}
+
+__NONSECURE_ENTRY
+void rtc_free(void)
+{
+    CLK_DisableModuleClock_S(rtc_modinit.clkidx);
+}
+
+__NONSECURE_ENTRY
+int rtc_isenabled(void)
+{
+    // NOTE: To access (RTC) registers, clock must be enabled first.
+    if (! (CLK->APBCLK0 & CLK_APBCLK0_RTCCKEN_Msk)) {
+        // Enable IP clock
+        CLK_EnableModuleClock_S(rtc_modinit.clkidx);
+    }
+
+    RTC_T *rtc_base = (RTC_T *) NU_MODBASE(rtc_modinit.modname);
+    
+    // NOTE: Check RTC Init Active flag to support crossing reset cycle.
+    return !! (rtc_base->INIT & RTC_INIT_ACTIVE_Msk);
+}
+
+__NONSECURE_ENTRY
+time_t rtc_read(void)
+{
+    /* NOTE: After boot, RTC time registers are not synced immediately, about 1 sec latency.
+     *       RTC time got (through RTC_GetDateAndTime()) in this sec would be last-synced and incorrect.
+     *       NUC472/M453: Known issue
+     *       M487: Fixed
+     */
+    if (! rtc_isenabled()) {
+        rtc_init();
+    }
+
+    /* Used for intermediary between date time of H/W RTC and POSIX time */
+    struct tm datetime_tm;
+
+    if (! t_hwrtc_origin_inited) {
+        t_hwrtc_origin_inited = 1;
+        
+        /* Convert date time from H/W RTC to struct TM */     
+        rtc_convert_datetime_hwrtc_to_tm(&datetime_tm, &DATETIME_HWRTC_ORIGIN);
+        /* Convert date time of struct TM to POSIX time */
+        if (! _rtc_maketime(&datetime_tm, &t_hwrtc_origin, RTC_FULL_LEAP_YEAR_SUPPORT)) {
+            return 0;
+        }
+
+        /* Load t_write from RTC spare register to cross reset cycle */
+        RTC_T *rtc_base = (RTC_T *) NU_MODBASE(rtc_modinit.modname);
+        RTC_WaitAccessEnable();
+        RTC_EnableSpareAccess();
+        RTC_WaitAccessEnable();
+        t_write = RTC_READ_SPARE_REGISTER(rtc_base, 0);
+    }
+
+    S_RTC_TIME_DATA_T hwrtc_datetime_2K_present;
+    RTC_WaitAccessEnable();
+    RTC_GetDateAndTime(&hwrtc_datetime_2K_present);
+    /* Convert date time from H/W RTC to struct TM */
+    rtc_convert_datetime_hwrtc_to_tm(&datetime_tm, &hwrtc_datetime_2K_present);
+    /* Convert date time of struct TM to POSIX time */
+    time_t t_hwrtc_elapsed;
+    if (! _rtc_maketime(&datetime_tm, &t_hwrtc_elapsed, RTC_FULL_LEAP_YEAR_SUPPORT)) {
+        return 0;
+    }
+
+    /* Present time in POSIX time */
+    time_t t_present = t_write + (t_hwrtc_elapsed - t_hwrtc_origin);
+    return t_present;
+}
+
+__NONSECURE_ENTRY
+void rtc_write(time_t t)
```

Comment:
It is not recommended to have secure veneer argument and return types as standard library defines, as they vary with compilers

Reply:
@deepakvenugopal I fix it in https://github.com/ARMmbed/mbed-os/pull/7302/commits/e61c5146c6131f60a70e253631860d4f2ffbb534.